### PR TITLE
feature/start-protocol-command

### LINKED ISSUE #207: Start protocol command line arguments
# Problem

Currently when starting a node, `relay`, `store`, and `filter` protocols are started automatically.

# Acceptance Criteria

There should be command line flags that allow a user to signal which protocols should be started.

### PR DESCRIPTION
**Pull request by decanus**
closes #207

### CODE REVIEW

**Comment by decanus** [comment]
this is sadly still buggy, in the tests. Need to figure out why

**Comment by decanus** [comment]
@oskarth @kdeme this works, but it seems a bit ugly. Any suggestions on cleanup?

File: waku/node/v2/config.nim [hunk 1]
```diff
@@ -73,6 +73,16 @@ type
       defaultValue: ""
       name: "storenode" }: string
 
+    store* {.
```
Comment (oskarth) [inline waku/node/v2/config.nim:76]: Do we want to have relay here too?

File: waku/node/v2/wakunode2.nim [hunk 2]
```diff
@@ -132,6 +119,21 @@ proc start*(node: WakuNode) {.async.} =
   ## XXX: this should be /ip4..., / stripped?
   info "Listening on", full = listenStr
 
+proc startFilter*(node: WakuNode) =
```
Comment (oskarth) [inline waku/node/v2/wakunode2.nim:122]: Maybe log here?

File: waku/node/v2/wakunode2.nim [hunk 3]
```diff
@@ -272,6 +277,12 @@ when isMainModule:
 
   waitFor node.start()
 
+  if conf.store:
+    startStore(node)
```
Comment (oskarth) [inline waku/node/v2/wakunode2.nim:281]: Is it starting it vs initing or mounting? Seems like latter to me

**Review by oskarth — APPROVED** [review]
In general lgtm, left some comments and suggestions

**Comment by oskarth** [comment]
> suggestions on clean up

I'd look at how NBC does it probably

File: waku/node/v2/config.nim (hunk 1 above)
Comment (decanus) [reply waku/node/v2/config.nim:76]: @oskarth probably

File: waku/node/v2/wakunode2.nim [hunk 4]
```diff
@@ -292,6 +299,15 @@ when isMainModule:
...
   waitFor node.start()
 
+  if conf.store:
+    mountStore(node)
+  
+  if conf.filter:
+    mountFilter(node)
+
+  if conf.filter:
+    await mountRelay(node)
```
Comment (kdeme) [inline waku/node/v2/wakunode2.nim:309]: You'll have to do waitFor here.

File: waku/node/v2/config.nim [hunk 5]
```diff
@@ -73,6 +73,21 @@ type
...
+      name: "store" }: bool
+
+    filter* {.
+      desc: "Flag whether to start filter protocol",
+      defaultValue: false
+      name: "filter" }: bool
+    
+    relay* {.
+      desc: "Flag whether to start relay protocol",
+      defaultValue: false
```
Comment (kdeme) [inline waku/node/v2/config.nim:88]: I guess this can be true? Or one of the protocols at least? Else nothing starts default?

**Review by kdeme — COMMENTED** [review]
I kinda had in mind not exposing this relay, filter and store (perhaps not the latter) directly to the user.
Sounds awfully difficult and I think the WakuNode abstraction could take care of this?

**Comment by decanus** [comment]
> I kinda had in mind not exposing this relay, filter and store (perhaps not the latter) directly to the user.
> Sounds awfully difficult and I think the WakuNode abstraction could take care of this?

@kdeme not sure what you are saying here.

**Comment by kdeme** [comment]
What I meant is that the application should probably have cli options in the trend of: `--full-node:true` or `--content-topics-only:blabla` (or `--light-node:true` if you want), `--historical-node:false`, etc. And then this could be mapped to the right protocols, which would be the role of the WakuNode abstraction imo. 
But ok, we haven't thought well about this yet, so I guess it is fine for testing for now.

**Comment by decanus** [comment]
@kdeme what do you think of the new flags?

File: waku/node/v2/config.nim [hunk 6]
```diff
@@ -73,6 +73,21 @@ type
       defaultValue: ""
       name: "storenode" }: string
 
+    historicalnode* {.
```
Comment (oskarth) [inline waku/node/v2/config.nim:76]: Sorry but these three ones seems worse to me, the previous one was simplistic but clear. Full node eg seems like a misnormer.

I'd make sure we know what we are doing and what we want before we increase the sophistication. KISS. Make sure we understand the problem well enough before writing a solution for it. See https://github.com/vacp2p/specs/issues/167 for context

**Comment by oskarth** [comment]
> What I meant is that the application should probably have cli options in the trend of: --full-node:true or --content-topics-only:blabla (or --light-node:true if you want), --historical-node:false, etc. And then this could be mapped to the right protocols, which would be the role of the WakuNode abstraction imo.
But ok, we haven't thought well about this yet, so I guess it is fine for testing for now.

Full node seems off to me. What does that mean for protocols? Not obvious. Content topics - boolean or list? For client or for receiver? Light node in what context? Historical node as server? On what topic?

What behaviour should be done for each of these aside from mounting the protocol?

All of this should be discussed at the problem formulation state, not in a PR imo.

**Comment by kdeme** [comment]
I was merely giving these as example to show that the current solution is not fine for the end user, not to actually copy over and use as is. Hence the "in the trend of". 
And I also stated that I'm fine with merging this as is, as we haven't resolved that issue.

File: waku/node/v2/wakunode2.nim [hunk 7]
```diff
@@ -125,20 +128,21 @@ proc start*(node: WakuNode) {.async.} =
   node.switch.mount(node.wakuFilter)
   node.subscriptions.subscribe(WakuFilterCodec, node.wakuFilter.subscription())
 
+proc mountStore*(node: WakuNode) =
+  info "mounting store"
+  node.wakuStore = WakuStore.init(node.switch, node.rng)
+  node.switch.mount(node.wakuStore)
+  node.subscriptions.subscribe(WakuStoreCodec, node.wakuStore.subscription())
+
+proc mountRelay*(node: WakuNode) {.async, gcsafe.} =
```
Comment (kdeme) [inline waku/node/v2/wakunode2.nim:137]: In mountRelay nothing gets mounted, as this happens in `init` already. 
Why this difference with the other protocols exactly? Could it be made consistent? Or would there be a downside to that?

Or could all get mounted in `init`, and then just run subscriptions/subscribe per protocol to activate it?
What would that do on a libp2p level? How would the connections look at after that `init` state then?

**Comment by decanus** [comment]
ok so let me revert to the previous.

**Comment by oskarth** [comment]
What's the state of this PR?

**Comment by decanus** [comment]
@oskarth theres some weird issues with the PR, I am working on fixing them.

**Comment by decanus** [comment]
@oskarth fixed. ptal

**Comment by decanus** [comment]
@oskarth should we modify the docker file to start the clusternode as a `wakunode2` running all 3 protocols?

**Comment by oskarth** [comment]
> @oskarth should we modify the docker file to start the clusternode as a wakunode2 running all 3 protocols?

This is the current behaviour right? And yeah we should

**Comment by oskarth** [comment]
Why is this PR touching nim-libp2p etc? Is it missing a rebase? I pushed a recent bump of submodules so I don't think we need one in this one.

File: waku/node/v2/wakunode2.nim [hunk 8]
```diff
@@ -168,8 +124,11 @@ proc subscribe*(node: WakuNode, request: FilterRequest, handler: ContentFilterHa
   ## Status: Implemented.
   info "subscribe content", filter=request
 
-  # @TODO: ERROR HANDLING
-  let id = await node.wakuFilter.subscribe(request)
+  # @TODO: GENERATE
```
Comment (oskarth) [inline waku/node/v2/wakunode2.nim:127]: Why is this a TODO? It looks like it is doing it

File: waku/node/v2/wakunode2.nim [hunk 9]
```diff
@@ -223,6 +182,56 @@ proc info*(node: WakuNode): WakuInfo =
...
+  node.switch.mount(node.wakuFilter)
+  node.subscriptions.subscribe(WakuFilterCodec, node.wakuFilter.subscription())
+
+proc mountStore*(node: WakuNode) =
+  info "mounting store"
+  node.wakuStore = WakuStore.init(node.switch, node.rng)
+  node.switch.mount(node.wakuStore)
+  node.subscriptions.subscribe(WakuStoreCodec, node.wakuStore.subscription())
+
+proc mountRelay*(node: WakuNode, topics: seq[string] = newSeq[string]()) {.async, gcsafe.} =
```
Comment (oskarth) [inline waku/node/v2/wakunode2.nim:202]: Out of curiosity, does it work when not mounting relay?

File: waku/node/v2/wakunode2.nim [hunk 10]
```diff
@@ -223,6 +182,56 @@ proc info*(node: WakuNode): WakuInfo =
...
+  await sleepAsync(5.seconds)
+
+  info "mounting relay"
+  proc relayHandler(topic: string, data: seq[byte]) {.async, gcsafe.} =
+    let msg = WakuMessage.init(data)
+    if msg.isOk():
+      node.filters.notify(msg.value(), "")
+      await node.subscriptions.notify(topic, msg.value())
+
+  await node.wakuRelay.subscribe("waku", relayHandler)
```
Comment (oskarth) [inline waku/node/v2/wakunode2.nim:224]: How does this interact with the other PR? Are they supposed to be merged in a specific order? I see later that we have mountRelay with topics as options, so why not use that argument?

File: waku/node/v2/wakunode2.nim [hunk 11]
```diff
@@ -300,10 +309,19 @@ when isMainModule:
...
 
   waitFor node.start()
 
+  if conf.store:
+    mountStore(node)
+  
+  if conf.filter:
+    mountFilter(node)
+
+  if conf.filter:
```
Comment (oskarth) [inline waku/node/v2/wakunode2.nim:322]:
```suggestion
  if conf.relay:
```

Typo

**Review by oskarth — CHANGES_REQUESTED** [review]

File: waku/node/v2/wakunode2.nim (hunk 9 above)
Comment (decanus) [reply waku/node/v2/wakunode2.nim:201]: It does not seem to work right now, I can investigate further why not.

File: waku/node/v2/wakunode2.nim (hunk 10 above)
Comment (decanus) [reply waku/node/v2/wakunode2.nim:223]: It doesn't really matter how it is merged as long as the five second sleep is there before subscribing from my understanding.

File: waku/node/v2/wakunode2.nim (hunk 9 above)
Comment (decanus) [reply waku/node/v2/wakunode2.nim:201]: @oskarth so I started 3 nodes:
 A) One with Relay, Filter and Store
 B) One with Relay
 C) One with Store

I sent a message from B, it was received by A, I then queried from C and received the expected message. So I would say it does work.

**Review by oskarth — CHANGES_REQUESTED** [review]
Previous feedback not addressed: vendor removal, dockerfile, (possibly more?).

Please comment on previous feedback with ack or so so it is clear if you have resolved it.

File: waku/node/v2/config.nim (hunk 5 above)
Comment (oskarth) [reply waku/node/v2/config.nim:88]: This is still false.

**Comment by decanus** [comment]
Alright @oskarth, docker has been addressed as have gitsubmodules and default option.

**Review by oskarth — APPROVED** [review]

**Comment by decanus** [comment]
@kdeme ptal

File: waku/node/v2/wakunode2.nim [hunk 12]
```diff
@@ -223,6 +181,56 @@ proc info*(node: WakuNode): WakuInfo =
...
+    #msgIdProvider = msgIdProvider,
+    triggerSelf = true,
+    sign = false,
+    verifySignature = false
+  )
+
+  node.wakuRelay = wakuRelay
+  node.switch.mount(wakuRelay)
+
+  await sleepAsync(5.seconds)
```
Comment (oskarth) [inline waku/node/v2/wakunode2.nim:214]: This was added here

**Pull request merged**